ironfish/src/blockchain/blockchain.ts

```diff
@@ -880,7 +881,7 @@ export class Blockchain {
         previousSequence = 0
         target = Target.maxTarget()
       } else {
-        const heaviestHead = this.head
+        const heaviestHead = headBlock ? headBlock.header : this.head
```

Comment:
Probably `head` was modified again before calling this code.

Reply:
Do you have a way to reproduce this issue consistently? I think it would be better to fix the issue by fetching `head` from the database instead of using `this.head`, like this:

```
        const head = await this.meta.get('head', tx)
        Assert.isNotUndefined(head)
        const heaviestHead = await this.getHeader(head, tx)
        Assert.isNotNull(heaviestHead)
```
Could you verify if that change fixes the issue as well?

Reply:
Actually, can you provide more logs from the error you're receiving?

It looks like that error would have come from `isValidAgainstPrevious`, which gets called while adding the block to the chain, and at that point it would have already verified that the `previousBlockHash` is correct. I'm not sure how in `newBlock` the previousBlockHash could be set to `heaviestHead.hash`, but then the target generated by `calculateTarget` is incorrect.

Reply:
This function can only ever calculate a block on the current head inside the lock, and no other block. The reason is because the note commitments need to match, but if the head changes and the note commitments are calculated they'll be wrong and not based on the passed in block. You could guarantee it by locking the chain DB when generating the block template, but I think it's better to probably return the head it's based on in the template result.